src/cfi.rs

```diff
@@ -89,6 +90,149 @@ impl<R: Reader> From<R> for DebugFrame<R> {
     }
 }
 
+/// `EhFrameHdr` contains the information about the `.eh_frame` section.
+///
+/// A pointer to the start of the `.eh_frame` data, and optionally, a binary
+/// search table of pointers to the `.eh_frame` records are found in this section.
+#[derive(Clone, Copy, Debug, PartialEq, Eq)]
+pub struct EhFrameHdr<R: Reader>(R);
+
+/// `ParsedEhFrameHdr` contains the parsed information fron the `.eh_frame` section.
+#[derive(Clone, Debug)]
+pub struct ParsedEhFrameHdr<R: Reader> {
+    addr_size: u8,
+    section: R,
+
+    eh_frame_ptr: Pointer,
+    fde_count: u64,
+    table_enc: DwEhPe,
+    table: R,
+}
+
+impl<'input, Endian> EhFrameHdr<EndianBuf<'input, Endian>>
+where
+    Endian: Endianity,
+{
+    /// Constructs a new `EhFrameHdr` instance from the data in the `.eh_frame_hdr` section.
+    pub fn new(section: &'input [u8], endian: Endian) -> Self {
+        Self::from(EndianBuf::new(section, endian))
+    }
+}
+
+impl<R: Reader> EhFrameHdr<R> {
+    /// Parses this `EhFrameHdr` to a `ParsedEhFrameHdr`.
+    pub fn parse(&self, bases: &BaseAddresses, addr_size: u8) -> Result<ParsedEhFrameHdr<R>> {
+        let mut reader = self.0.clone();
+        let version = reader.read_u8()?;
+        assert_eq!(version, 1); // FIXME return error
+
+        let eh_frame_ptr_enc = DwEhPe(reader.read_u8()?);
+        let fde_count_enc = DwEhPe(reader.read_u8()?);
+        let table_enc = DwEhPe(reader.read_u8()?);
+        let eh_frame_ptr = parse_encoded_pointer(eh_frame_ptr_enc, bases, addr_size, &self.0, &mut reader)?;
+        let fde_count = parse_encoded_pointer(fde_count_enc, bases, addr_size, &self.0, &mut reader)?;
+        let fde_count = match fde_count {
+            Pointer::Direct(c) => c,
+            Pointer::Indirect(_) => unreachable!(),
```

Comment:
Is this really unreachable? I would expect there to be some early return if `fde_count_enc.is_indirect()` in order to make this actually unreachable, and not just a bad `.eh_frame_hdr`.

Reply:
I think this probably should just return an error.

I mean, this isn't even a pointer, it's just a count that uses the pointer-encoding mechanism. I can't imagine any sane (or insane) reason why a compiler would emit an indirect pointer here and supporting this case just makes everything so very awkward (since we definitely can't just deref) - so why bother when this won't ever happen?

Reply:
> I think this probably should just return an error.

Agreed, and we should have test coverage.